aether-ui/aether/ui/assets/apps/pipeline/components/ContractAddButton.jsx

```diff
@@ -22,15 +22,24 @@ import React, { Component } from 'react'
 import { connect } from 'react-redux'
 import { FormattedMessage } from 'react-intl'
 
-import { addContract } from '../redux'
+import { addContract, selectPipeline } from '../redux'
 
 class ContractAddButton extends Component {
+  createNewContract () {
+    const { id } = this.props.pipeline
+    this.props.selectPipeline(id)
+    this.props.history.push({
+      pathname: `/${id}`,
+      state: { isNewContract: true }
+    })
+  }
+
```

Comment:
```javascript
class ContractAddButton extends Component {
  createNewContract () {
    const { id } = this.props.pipeline
    this.props.addBlankContract(id)
  }
  // ...
}
```

```javascript

// -----------------------------------------------------------------------------
// redux.jsx
// -----------------------------------------------------------------------------
export const types = {
  // ...
  PIPELINE_ADD_CONTRACT: 'pipeline.add.contract'
}

export const addBlankContract = (pid) => ({
  type: types.PIPELINE_ADD_CONTRACT,
  payload: pid
})


const reducer = (state = INITIAL_STATE, action) => {
  const nextState = { ...state, ...ACTIONS_INITIAL_STATE }

  switch (action.type) {
    // ...
    case types.PIPELINE_ADD_CONTRACT: {
      // if the add contract button was pressed in the pipelines list page
      // pipelineList contains all pipelines otherwise it's empty and the
      // current pipeline is the indicated one
      const currentPipeline = state.pipelineList.find(p => p.id === action.payload) || state.currentPipeline
      const currentContract = { 
        id: 'new-contract' // this is an example of implementing this
      }

      return {
        ...nextState,
        currentSection: CONTRACT_SECTION_SETTINGS,
        currentPipeline,
        currentContract
      }
    }

    case types.PIPELINE_BY_ID: {
      const currentPipeline = parsePipeline(action.payload)
      // if selected contract id is `new-contract` this means that the
      // add contract button was pressed in the pipeline list page
      // the selected contract is not in the list "yet".
      const currentContract = (state.currentContract.id === 'new-contract'
        ? currentContract
        : findContract(currentPipeline, state.currentContract && state.currentContract.id)
      )

      return {
        ...nextState,
        pipelineList: replaceItemInList(state.pipelineList, currentPipeline),
        currentPipeline,
        currentContract
      }
    }
  }
}
```
With this (more or less), no changes are required in the `Pipeline` component and in the `Settings` one you only need to choose between the `addContract` or `updateContract` action depending on the `id`.
If you press "cancel" in the settings button just remove the "blank" contract and take the first one of the list or...

@adamvert this is what I mean with simplify and keep clean the components.

Reply:
On a second look, the act of adding a new contract is useable only on one component `Pipeline.jsx`, thus i don't see the need to redux-ify it. What is needed is the ability to switch the `currentContract` value with a new contract object which is handled by the `contractChanged` action. NOTE: `addContract` and `updateContract` have server-side side effects. The point is to delay the contract creation